README.md
# Gridsome source MySQL

**Alternate version with image cdn support using [cloudinary](https://cloudinary.com/invites/lpov9zyyucivvxsnalc5/n4iiwbfiyryrdnlqkrye) and [jsonbin.io](https://jsonbin.io)**

Gridsome Source Plugin to load data directly from MySQL Database

  * If you don't succeed through a CMS, bypass it and load the data directly from the MySQL Database

  * Specify names of columns containing image urls to download and optimized them with Gridsome. Supports single and comma delimited image urls.

  * Build up your mysql queries as you need to get the data you require.

  * Supports sub queries and references between content types

  * `id` field from mysql is renamed to `mysqlId`

  * Content type references via `xxx_id`(single id) and `xxx_ids`(comma delimited ids)

  * Cloudinary images also have placeholders which will be generated and saved to jsonstore for future builds.


View the [changelog](https://github.com/u12206050/gridsome-source-mysql/blob/master/CHANGELOG.md) for any possible changes from previous versions.

### Latest Updates

  *2.7.0* BREAKING CHANGE: Switched from jsonstore to jsonbin

  *2.6.0* Added optimsed for cloudinary g-image

## Install

  `npm install gridsome-source-mysql-cloudinary --save`

## Setup

> Make sure your mysql database is accessible everywhere you are planning to build your site from.

### Cloudinary (for images)

Create a FREE account on [cloudinary](https://cloudinary.com/invites/lpov9zyyucivvxsnalc5/n4iiwbfiyryrdnlqkrye) and then enable auto uploading by doing the following:

  1. Navigate to [settings > upload](https://cloudinary.com/console/settings/upload)
  2. Then under `Auto upload mapping` in the `Folder` field add the name of the directory where all your images are. They may be in subdirectories as well. eg. `media`
  3. Add the the full url that includes the folder name in the `URL prefix` field eg. `https://example.no/media/`
  4. In the config (see below) update the `name`, `folder` and `match` fields.
  5. Update all your queries where you query for images to include the srcset and other required fields. (See example **Usage** below)

### Use the optimized CloudinaryImage as `g-image`

In `main.js` import and override Gridsome's image component with CloudinaryImage. As a drop-in-replacement we will call it `g-image` so your code does not need to be updated.

```
// main.js

import CloudinaryImage from 'gridsome-source-mysql-cloudinary/CloudinaryImage'

export default function(Vue) {

  /* Drop-in-replacement for Gridsome's image */
  Vue.component('g-image', CloudinaryImage)

}
```

### [Jsonbin](https://jsonbin.io)

Create a free account which should be enough transactions if you aren't building too often with too many images.
Update the config as specified below with you jsonbin key, binId (which you create before hand), and optional collectionId to group all your image bins.

### Config

Within plugins in the `gridsome-config.js` file, add the connection settings and queries for the data you need.

```javascript:title=gridsome-config.js
// gridsome-config.js

module.exports = {
  plugins: [
    {
      use: 'gridsome-source-mysql-cloudinary',
      options: {
        connection: {
          host: 'localhost', // required
          port: 3306,
          user: 'root', // required
          password: 'secret', // required
          database: 'my_db', // required
          connectionLimit : 10
        },
        debug: true, // Default false on production
        ignoreImages: false, // Do not process any images
        jsonbin: { // Create a free account on [jsonbin.io](https://jsonbin.io)
          key: process.env.JSONBIN_KEY, // Secret key from [jsonbin.io](https://jsonbin.io/api-keys)
          binId: process.env.JSONBIN_BIN, // The id of the bin to save data too
          collectionId: process.env.JSONBIN_COLLECTION, // (optional) The id of the collection in which to save additional bins
        },
        cloudinary: {
          name: 'example',
          folder: 'media',
          uri: 'c_scale,e_vectorize,w_50', // Scaling for svg placeholder
          sizes: ['480', '800'],
          match: /https?:\/\/(www\.)?example\.no\/media\// // Url to match for images and swap out for the cloudinary url
        },
        regex: /()_\d(.(jpg|png|svg|jpeg))/i, // Default false
        queries: [ // required
          {
            name: 'Author',
            route: '/authors/:path',
            path: 'fullname',
            sql: `SELECT id, fullname, avatar, url FROM author`,
            images: ['avatar'] // Default []
          },
          {
            name: 'Post',
            path: 'title',
            sql: `SELECT id, title, image, gallery, author as author_id, excerpt, body, created FROM post WHERE published = ?`,
            args: [1],
            images: ['image', ['gallery']] //Gallery contains comma delimited string of image url.
          }
        ]
      }
    }
  ]
}
```

Relationship ids should be in the format of `xxx_id` where `xxx` is the name of another query.

## Usage

On the above example two content types will be created `Post` and `Author` with `author_id` being a relation:

```
query {
  allPost {
    edges {
      node {
        title
        path
        image {
          src
          dataUri
          srcset
          size {
            width
            height
          }
        }
        gallery {
          index
          image {
            src
            dataUri
            srcset
            size {
              width
              height
            }
          }
        }
        excerpt
        author {
          fullname
          url
          image {
            src
            dataUri
            srcset
            size {
              width
              height
            }
          }
        }
      }
    }
  }
}
```

## Definitions

### Options

  *regex*: Specify false to not use or a `regex` expression that has 2 capture groups. This can be used to remove duplicate files for example the value `/()_\d(.(jpg|png|svg|jpeg))/i` renames all files that end with `_\d` eg. `_1`, `_2`; since we assume them to be duplicate files. We DO NOT change the source url since it might be the original file (not ending with `_\d`) isn't used.

### Query

Field | Type | Info
---|---|---
name | string | Name of the resulting content type
route? | string | Specify a [dynamic route](https://gridsome.org/docs/routing) structure eg. `/blog/:path`
path | function(slugify, row, parentRow?): string | Return the path for the given row
path | { prefix?: string, field: string, suffix?: string } | `field` should exist on each row and will be slugified
path | string | Name of a field on each row to slugify and use as path
sql | string | A SQL Query with optional placeholders `?` which will be replaced by args in order
args? | array<string> | Simple array of static values
args? | function(parentRow?): array<string> | Return array of values based on data from the parentRow or dynamic calculated data, eg. [Date.now()]
json | array<string> | Names of fields containing json to be converted to objects else `null`
images? | array<string|string[]|function(row, addImageUrl(url: string))> | Names of fields on rows that contain urls of images to download and optimize via Gridsome
subs | array<Query> | Array of Query to execute per result of the current query

# MySQL Query Examples

## Generate comma seperated urls and ids

The following is an example of how you can generated the fields for using as a `one-to-many` relationship in graphql and also joining image urls.

```
queries: [
  {
    name: 'Product',
    path: 'slug',
    images: ['image', ['gallery']] // Default []
    sql: `SELECT
      pc.product_id as id,
      cats.category_ids,
      pc.sku,
      pc.name,
      pc.price,
      pc.slug,
      CONCAT('https://example.com/media/', pc.image) as image,
      media.images as 'gallery',
      FROM product_catalog pc
      INNER JOIN (
          SELECT product_id, GROUP_CONCAT(CONCAT('https://example.com/media/catalog/product',value)) AS 'images'
          FROM product_media
          GROUP BY product_id
        ) media
      ON media.product_id = pc.product_id
      INNER JOIN (
          SELECT product_id, GROUP_CONCAT(category_id) AS 'category_ids'
          FROM product_category
          GROUP BY product_id
        ) cats
      ON cats.product_id = pc.product_id
      WHERE pc.status = 1`
  },
  {
    name: 'Category',
    route: '/category/:path',
    path: 'path',
    images: ['image'],
    sql: `SELECT
      category_id AS 'id',
      name,
      CONCAT('https://example.com/media/', image) as image,
      parent_id AS 'category_id',
      position,
      level,
      product_count AS 'count'
      FROM category_catelog
      WHERE active = 1`
  }
]
```

In the above example `cats.category_ids` will result in an array of `Category` content types if you have specified a query for `Category`

Images in this database were relative, so in order for them to be downloaded they need to be concatenated with the site origin.

In the `Category` query, we change the `parent_id` to output as `category_id` since we want it to be linked to another `Category` automatically.


### The drop-in-replacement, CloudinaryImage

This drop-in version will filter through the appropriate sizes available, eg. `sizes: ['480', '800', '1200']`, on each image, such that even though larger images exist it will only use the image that is `>= width`, width being speicified on the tag.

Example: `<g-image :src="image" width="600">` here the image with the generated size of `800` will be used, and not `1200`.

CHANGELOG.md
2.7.0 [BREAKING] Changed from jsonstore to jsonbin.io format
  * Requires creating free account with jsonbin.io and updating parameters.

2.6.0 Added optimsed for cloudinary g-image

  * Component to replace Gridsome's image in order to better use images from Cloudinary

2.5.0 Support for parsing json

  * Version bump to match mysql-plugin

  * Support callback function in images.

2.0.0 Alternate version including cloudinary

  * Store images on cloudinary with support for placeholders and dynamic resizing etc.

1.4.10 Removed support for GIF, since it breaks when trying to optimize

1.4.7 Added regex option to clean up image file names

  * Useful for removing duplicate images that are the same.

1.4.5 Support for dynamic routes added

  * Now you can make use of dynamic routes by specifying a `route` for your query. This saves space in the final `routes.js` file eg:

  **Before**
  ```
  {
    name: 'Author',
    path: {
      prefix: '/author',
      field: 'name'
    },
    …
  }
  ```

  **After**
  ```
  {
    name: 'Author',
    route: '/author/:path',
    path: 'name',
    …
  }
  ```

  * Read more about Gridsome routing [here](https://gridsome.org/docs/routing)

1.4.3 Added back status output for downloading files

1.4.2 Fixed version 1.4.0

1.4.0 *Broken Version* Swapped out queue for [p-Map](https://github.com/sindresorhus/p-map)

1.3.3 Set download concurrency to no of cpus

1.3.2 Check relation name

  * Implemented a check for a query with the name `xxx` from `xxx_id` before creating a relation.

1.3.1 Changed `_id` field containing original id to `mysqlId`

  * Unpublished version 1.3.0

1.3.0 *Broken Version* Support for array of image urls. Keep original ID

  * Within the image options field for each query you can specify an array with the one and only element being the field name of the field containing a comma delimited string of image urls to be downloaded.

  * The original id value if one exists will now be available as `mysqlId`

1.2.2 Support for string array of ids

  * Fields in the form of `xxx_ids` will now be parsed as a comma delimited string of ids that will be split up and used to create relationships of.

1.2.0 Added download queue

  * When downloading a lot of images, it helps having a queue to regulate how many files to download in parallel. Currently setting to 3

1.1.0 Fixed ID Field

file.js
/* Usage */
/*
const file = require('./file.js')

const imageDirectory = 'tmp_images'
const images = {}
let loadImages = false

// Somewhere on each possible image url
if (url && String(url).match(/^https:\/\/.*\/.*\.(jpg|png|svg|jpeg)($|\?)/i)) {
  const filename = file.getFilename(url)
  const id = makeUid(url)
  const filepath = file.getFullPath(imageDirectory, filename)
  if (!images[id]) images[id] = {
    filename,
    url,
    filepath
  }

  loadImages = true

  // Assign filepath as the new value for the image
  obj[imgField] = filepath
}

// After processing all fields
if (loadImages) await downloadImages(images)

async function downloadImages(images) {
  file.createDirectory(imageDirectory)

  let exists = 0
  const download = []
  Object.keys(images).forEach(async (id) => {
    const { filename, filepath } = images[id]

    if (!file.exists(filepath)) {
      download.push(images[id])
    } else exists++
  })

  ISDEV && console.log(`${exists} images already exists and ${download.length} to download`)

  if (download.length) {
    await pMap(download, async ({ filename, url, filepath }) => {
      await file.download(url, filepath)
      ISDEV && console.log(`Downloaded ${filename}`)
    }, {
      concurrency: os.cpus * 2
    })
  }
}
*/

const https = require('https')
const fs = require('fs-extra')
const path = require('path')

const TMPDIR = '.temp/downloads'
let tmpCount = 0

function createDirectory(dir) {
  const pwd = path.resolve(dir)
  if (!fs.existsSync(pwd)) fs.mkdirs(pwd)

  return pwd
}

createDirectory(TMPDIR)

function getFullPath(dir, filename) {
  return path.resolve(dir, filename)
}

function getFilename(url, regex) {
  let name = url.replace(/%2F/g, '/').split('/').pop().replace(/\#(.*?)$/, '').replace(/\?(.*?)$/, '')
  return regex ? name.replace(regex, '$1$2') : name
}

function exists(filepath) {
  return fs.existsSync(filepath)
}

function downloadIfNotExists(url, absPath) {
  if (!url ||  !path || exists(absPath)) return

  return download(url, absPath)
}

function download(url, absPath) {
  return new Promise(function(resolve) {
    const tmpPath = getFullPath(TMPDIR, `${++tmpCount}.tmp`)
    const file = fs.createWriteStream(tmpPath)
    const request = https.get(url, (response) => {
      response.pipe(file)
      file.on('finish', () => {
        file.close()
        fs.rename(tmpPath, absPath, resolve)
      })
    }).on('error', (err) => {
      console.error(err.message)
      fs.unlink(String(tmpPath), resolve)
    })
  })
}

module.exports = {
  createDirectory,
  getFullPath,
  getFilename,
  exists,
  download
}

CloudinaryImage.js
// @vue/component
export default {
  name: 'g-image',
  functional: true,
  props: {
    src: { type: [Object, String], required: true },
    width: { type: String, default: '' },
    height: { type: String, default: '' },
    quality: { type: String, default: '' },
    fit: { type: String, default: '' },
    position: { type: String, default: '' },
    background: { type: String, default: '' },
    blur: { type: String, default: '' },
    immediate: { type: true, default: undefined }
  },

  render: (h, { data, props }) => {
    const classNames = [data.class, 'g-image']
    const noscriptClassNames = [data.staticClass, classNames.slice()]
    const isImmediate = props.immediate || props.immediate !== undefined
    const directives = data.directives || []
    const attrs = data.attrs || {}
    const hook = data.hook || {}
    const res = []

    if (props.src === null) {
      console.error('Image src prop can not be null')
      return null
    }

    switch (typeof props.src) {
      case 'string':
        attrs.src = props.src
        break

      case 'object': {
        const { src, srcset, sizes, size, dataUri } = props.src
        const isLazy = !isImmediate && dataUri

        attrs.src = isLazy ? dataUri : src
        attrs.width = size.width

        if (isLazy) attrs['data-src'] = src
        if (srcset.length) {
          const { width } = props
          if (width) {
            let newSrc = null
            if (width <= 100 && src.indexOf('image/upload/') > 0) {
              newSrc = src.replace(/image\/upload\/[^\/]+\//, 'image/upload/c_scale,w_100,h_100/')
              attrs[`${isLazy ? 'data-' : ''}srcset`] = `${newSrc} 100w`
              attrs['data-src'] = newSrc
            } else {
              let newSrcSet = []

              srcset.forEach(s => {
                let parts = s.split(' ')
                let set = {
                  w: parts[1].replace('w',''),
                  src: parts[0]
                }
                if (set.w >= width && (!newSrc || set.w <= newSrc.w)) newSrc = set
                if (set.w < width) newSrcSet.push(`${set.src} ${set.w}w`)
              })

              if (newSrc) newSrcSet.push(`${newSrc.src} ${newSrc.w}w`)
              else newSrc = newSrcSet[newSrcSet.length - 1]

              attrs[`${isLazy ? 'data-' : ''}srcset`] = newSrcSet.join(', ')
              attrs['data-src'] = newSrc.src
            }
          } else {
            attrs[`${isLazy ? 'data-' : ''}srcset`] = srcset.join(', ')
          }
        }
        if (sizes) attrs[`${isLazy ? 'data-' : ''}sizes`] = sizes

        if (isLazy) {
          directives.push({ name: 'g-image' })
        }

        break
      }
    }

    hook.update = (oldVnode, vnode) => {
      const { attrs: oldAttrs = {}} = oldVnode.data
      const { attrs = {}} = vnode.data

      if (attrs['data-src'] && attrs.src !== oldAttrs.src) {
        // clear srcset and sizes to show the dataUri image
        vnode.elm.srcset = ''
        vnode.elm.sizes = ''
      }
    }

    res.push(h('img', {
      ...data,
      class: classNames,
      directives,
      props,
      attrs,
      hook
    }))

    if (attrs['data-src']) {
      classNames.push('g-image--lazy')
      classNames.push('g-image--loading')
      noscriptClassNames.push('g-image--loaded')

      // must render as innerHTML to make hydration work

      res.push(h('noscript', {
        domProps: {
          innerHTML: `` +
            `<img src="${props.src.src}" class="${noscriptClassNames.join(' ')}"` +
            (attrs.width ? ` width="${attrs.width}"`: '') +
            (attrs.alt ? ` alt="${attrs.alt}"` : '') +
            `>`
        }
      }))
    }

    return res
  }
}

index.js
const os = require('os')
const readline = require('readline')
const mysql = require('mysql')
const pMap = require('p-map')
const file = require('./file.js')

const probe = require('probe-image-size')
const imageDataURI = require('image-data-uri')

const axios = require('axios')

let cpus = os.cpus().length
cpus = cpus > 2 ? cpus : 2

let ISDEV = process.env.NODE_ENV === 'development'

/* Capitalize string */
function capitalize(str) {
  str = str.trim()
  return str.trim().charAt(0).toUpperCase() + str.slice(1)
}

class MySQLSource {
  static defaultOptions () {
    return {
      debug: false,
      ignoreImages: false,
      imageDirectory: 'sql_images',
      regex: false,
      queries: [],
      jsonId: false,
      jsonChunkSize: 20,
      connection: {
        host: 'localhost',
        port: 3306,
        user: 'root',
        password: 'secret',
        database: 'my_db',
        connectionLimit : 10
      }
    }
  }

  constructor (api, options = MySQLSource.defaultOptions()) {

    const opts = {
      ...MySQLSource.defaultOptions(),
      ...options
    }

    this.pool = mysql.createPool({
      ...opts.connection
    });

    ISDEV = opts.debug

    this.cTypes = {}
    this.paths = {}
    this.queries = opts.queries || []
    if (!this.queries.length) throw new Error('No queries to load')

    this.loadImages = false
    this.regex = opts.regex
    this.images = false
    this.imageDirectory = opts.imageDirectory

    if (opts.cloudinary) {
      const { name, match, folder, uri, sizes } = opts.cloudinary
      this.cloud = {
        name, match, folder, uri, sizes,
        isMatch: (url) => {
          return url.match(match)
        },
        getPath: (url) => {
          return url.replace(match, '')
        },
        toUrl: (path, size) => {
          return `https://res.cloudinary.com/${name}/image/upload${size ? `/${size}/` : '/'}${folder}/${path}`
        }
      }
    }

    /* SETUP JSONBin Axios connection */
    const jsonbin = (() => {
      if (!opts.jsonbin || !opts.jsonbin.key) return false
      const options = {
        baseURL: 'https://api.jsonbin.io/b/',
        headers: {
          'secret-key': opts.jsonbin.key,
          'collection-id': opts.jsonbin.collectionId,
          versioning: false
        }
      }
      if (opts.jsonbin.collectionId) options['collection-id'] = opts.jsonbin.collectionId

      return axios.create(options)
    })()

    api.loadSource(async (store) => {
      this.store = store

      let rdata, q, existingBinIds = [];
      if (!opts.ignoreImages) {
        this.images = {}
        if (jsonbin) await this.loadImageCache(opts, jsonbin)
      }

      this.checkQNames(this.queries)

      await this.fetchQueries(this.queries)
      this.pool.end(function(err) {
        ISDEV && console.log('MySQL Connections Closed')
      })

      if (this.images) {
        if (this.loadImages) await this.downloadImages()
        if (jsonbin) await this.saveImageCache(opts, jsonbin)
      }
    })
  }

  checkQNames(queries) {
    Array.isArray(queries) && queries.forEach((Q) => {
      Q.name = capitalize(Q.name)
      if (this.cTypes[Q.name]) console.warn(`You should not have two queries with the same name. ${Q.name}`)
      else this.cTypes[Q.name] = true
      if (Q.subs) this.checkQNames(Q.subs)
    })
  }

  async fetchQueries(queries, parentQuery, parentRow) {
    const { slugify, addContentType, makeUid, createReference } = this.store

    await Promise.all(queries.map(async (Q) => {
      const args = (typeof Q.args === 'function' ? Q.args(parentRow) : Q.args) || null
      const sql = mysql.format(Q.sql, args)

      const cType = this.cTypes[Q.name] = addContentType({
        typeName: Q.name,
        route: Q.route
      })

      const rels = []

      const rows = await new Promise((resolve, reject) => {
        this.pool.query(sql, (error, results, fields) => {
          if (error) throw new Error(error)

          /* Find relationship fields */
          let hasIdField = false
          for (const f in fields) {
            const field = fields[f].name
            hasIdField = field === 'id' || hasIdField
            const matches = field.match(/^(.+)_(ids?$)/)
            if (matches && matches.length > 2) {
              const qname = matches[1]
              const qtype = capitalize(qname)
              if (this.cTypes[qtype]) {
                rels.push({
                  type: qtype,
                  name: qname,
                  field,
                  isArray: matches[2] === 'ids'
                })
              } else {
                console.warn(`No query with name "${qname}" exists. Not creating relation`)
              }
            }
          }

          if (!hasIdField) throw new Error('Rows must have id field')

          resolve(results)
        })
      })

      if (!Array.isArray(rows)) rows = []

      console.log(`${Q.name}: retrieved ${rows.length} results`)

      let PathFn = Q.path
      if (typeof PathFn !== 'function') {
        /* Default path function */
        PathFn = (slugify, row, parent) => {
          let slug = `/${Q.name}/${row.id}`
          if (typeof Q.path === 'object') {
            slug = Q.path.prefix || ''
            slug += `/${row[Q.path.field] ? slugify(row[Q.path.field]) : row.id}`
            slug += Q.path.suffix || ''
          } else if (typeof Q.path === 'string' && row[Q.path]) {
            slug = slugify(row[Q.path]) || slug
          }
          return slug
        }
      }

      return Promise.all(rows.map(async (row, i) => {
        row.mysqlId = row.id
        row.id = makeUid(`${Q.name}–${row.id}`)
        row.path = PathFn(slugify, row, parentRow)

        if (this.paths[row.path]) {
          row.path = `${row.path}-${this.paths[row.path]++}`
        } else this.paths[row.path] = 1

        if (parentQuery && parentRow)
          row._parent = createReference(parentQuery.name, parentRow.id)

        /* Parse JSON fields */
        if (Array.isArray(Q.json)) {
          Q.json.forEach(jsonField => {
            try {
              row[jsonField] = JSON.parse(row[jsonField])
            } catch (e) {
              row[jsonField] = null
            }
          })
        }

        /* Check for images */
        if (this.images && Array.isArray(Q.images)) {
          await pMap(Q.images, async imgField => {
            if (typeof imgField === 'function') {
              await imgField(row, (url) => this.addImage(url))
            }
            if (Array.isArray(imgField)) {
              if (imgField.length !== 1) throw new Error('MySQL query image array should contain exactly 1 field')
              const imageUrls = String(row[imgField[0]]).split(',')
              row[imgField[0]] = await Promise.all(imageUrls.map(async (url, i) => ({
                index: i,
                image: await this.addImage(url)
              })))
            } else {
              row[imgField] = await this.addImage(row[imgField])
            }
          }, { concurrency: 2 })
        }

        /* Check for relationships */
        rels.forEach(rel => {
          if (rel.isArray) {
            const ids = String(row[rel.field]).split(',')
            row[rel.name] = ids.map(id => createReference(rel.type,
              makeUid(`${rel.type}–${id}`)))
          } else {
            row[rel.name] = createReference(rel.type, makeUid(`${rel.type}–${row[rel.field]}`))
          }
        })

        cType.addNode(row)

        /* Check sub queries to execute with parent */
        if (Array.isArray(Q.subs)) return this.fetchQueries(Q.subs, Q, row)

        return row
      }))
    }))
  }

  async addImage(url) {
    if (url && String(url).match(/^https:\/\/.*\/.*\.jpg|png|svg|jpeg($|\?)/i)) {
      const { images, regex, store, imageDirectory, cloud } = this
      const filename = file.getFilename(url, regex)
      const iId = store.makeUid(filename)
      const filepath = file.getFullPath(imageDirectory, filename)

      if (!images[iId]) {
        if (cloud) {
          if (!cloud.isMatch(url)) return null
          const path = cloud.getPath(url)
          const src = cloud.toUrl(path, `f_auto`)

          let meta, imageUri = null
          try {
            meta = await probe(cloud.toUrl(path), { retries: 4 })
          } catch(err) {
            console.warn(`Failed loading meta for ${src}`)
            return null
          }
          try {
            imageUri = await imageDataURI.encodeFromURL(cloud.toUrl(path, cloud.uri), {
              timeout: 20000
            })
          } catch(err) {
            console.warn(`Failed loading image uri for ${src}`)
            return null
          }

          const srcset = []
          cloud.sizes.forEach(size => {
            if (size < meta.width) {
              srcset.push(`${cloud.toUrl(path, `f_auto,c_limit,q_auto:best,w_${size}`)} ${size}w`)
            }
          })

          srcset.push(`${src} ${meta.width}w`)

          images[iId] = {
            src,
            srcset,
            sizes: `(max-width: ${meta.width}w) 100vw, ${meta.width}w`,
            name: filename,
            dataUri: `data:image/svg+xml,<svg fill='none' viewBox='0 0 800 800' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'><defs><filter id='__svg-blur'><feGaussianBlur in='SourceGraphic' stdDeviation='30'/></filter></defs><image x='0' y='0' filter='url(%23__svg-blur)' width='800' height='800' xlink:href='${imageUri}' /></svg>`,
            size: {
              width: meta.width,
              height: meta.height
            }
          }
        } else {
          images[iId] = {
            filename,
            url,
            filepath
          }
          this.loadImages = true
        }
      }

      return images[iId].filepath ? images[iId].filepath : images[iId]
    }
    return null
  }

  async downloadImages() {
    const { images, imageDirectory } = this
    file.createDirectory(imageDirectory)

    let exists = 0
    const download = []
    Object.keys(images).forEach(async (iId) => {
      const { src, filename, filepath } = images[iId]
      if (src) return

      if (!file.exists(filepath)) {
        download.push(images[iId])
      } else exists++
    })

    const total = download.length
    let progress = 0
    function status(msg) {
      readline.clearLine(process.stdout, 0)
      readline.cursorTo(process.stdout, 0, null)
      process.stdout.write(msg)
    }

    console.log(`${exists} images already exists with ${total} images to download`)

    if (total) {
      await pMap(download, async ({ filename, url, filepath }) => {
        await file.download(url, filepath)
        status(`${Math.round((++progress)*100/total)}% – Downloaded ${filename}`)
      }, {
        concurrency: cpus * 2
      })

      status('100% – ')
      console.log(`${total} images downloaded`)
    }

    this.loadImages = false
  }

  async loadImageCache(opts, jsonbin) {
    const { images } = this

    try {
      const binInfo = await jsonbin.get(opts.jsonbin.binId).then(res => res.data)
      if (binInfo && binInfo.count && binInfo.binIds) {
        this.existingBinIds = binInfo.binIds
        console.log(`Loading ${this.existingBinIds.length} chunks containg ${binInfo.count} images`)

        let loaded = 0
        await pMap(this.existingBinIds, async (bId, i) => {
          const binImages = await jsonbin.get(bId).then(res => res.data)
          Object.keys(binImages).forEach(iId => {
            if (typeof binImages[iId] === 'object') {
              images[iId] = binImages[iId]
              loaded++
            }
          })
        }, { concurrency: 4 })

        try {
          this.checksum = JSON.stringify(images).length
        } catch(err) {
          console.log('Failed creating checksum')
        }

        console.log(`Loaded ${loaded} images from jsonbin`)
      } else {
        console.log(`No images from jsonbin or invalid bin id`)
      }
    } catch(err) {
      this.images = {}
      console.log('Error loading from jsonbin')
      console.log(err.message)
    }
  }

  async saveImageCache(opts, jsonbin) {
    const { images, checksum, existingBinIds } = this
    try {
      if (checksum > 10 && checksum === JSON.stringify(images).length) {
        console.log('Checksum passed, no need to update image cache')
        return
      }
    } catch(err) {
      console.log('Failed checking checksum')
    }

    try {
      const iIds = Object.keys(images)
      const count = iIds.length

      const chunks = []
      let chunk = {}
      const chunkSize = 60
      let size = chunkSize

      iIds.forEach(iId => {
        if (size--) {
          chunk[iId] = this.images[iId]
        } else {
          chunks.push(chunk)
          size = chunkSize
          chunk = {}
        }
      })

      console.log(`Chunking and saving ${count} images`)

      if (size !== chunkSize) {
        chunks.push(chunk)
      }

      const usedBinIds = []
      await pMap(chunks, async (chunk, cI) => {
        let bId = existingBinIds.pop()
        if (bId) {
          return jsonbin.put(bId, chunk).then(res => {
            usedBinIds.push(bId)
            console.log(`${cI} Updated bin`)
            return cI
          }).catch((err) => {
            console.log(`${cI} failed: ${err.message}`)
          })
        } else {
          return jsonbin.post('', chunk).then(res => {
            usedBinIds.push(res.data.id)
            console.log(`${cI} Created new bin`)
            return jsonbin.put(opts.jsonbin.binId, {
              binIds: usedBinIds,
              count
            })
          }).catch((err) => {
            console.log(`${cI} failed: ${err.message}`)
          })
        }
      }, { concurrency: 4 })

      await jsonbin.put(opts.jsonbin.binId, {
        binIds: usedBinIds,
        count
      })

      /* Cleanup */
      await Promise.all(existingBinIds.map((bId) => jsonbin.delete(bId).then((res) => {
        console.log(`Deleted ${bId}`)
        return res.data
      }))).catch(() => {
        console.log("Don't panic :) FYI, I tried to delete unused bins but failed. But otherwise I have ")
      })

      console.log(`Saved ${count} images thumbnails to jsonbin`)

    } catch(error) {
      console.log('Error saving to jsonbin')
      console.log(error.message)
    }
  }
}

module.exports = MySQLSource